### ISSUE TITLE
How to show two structures are equal?

### ISSUE THREAD

**Issue by MitchRowett** [semorrison/2017-summer-students#35]
I've reached a stage where the goal is

`⊢ {to_subgroup := {set := {x : α | f.map x = 1}, one_closed := _, inv_closed := _, mul_closed := _}, normal := _} =
    {to_subgroup := {set := {1}, one_closed := _, inv_closed := _, mul_closed := _}, normal := _}`

I would assume that this could be reduced to the goal `⊢ {x : α | f.map x = 1} = {1}`, since this is the only information which matters. Is this possible?

**Comment by semorrison** [comment]
My copy of Lean seems to be broken at the moment. Is there a `congr_struct` tactic? I seem to remember using that to take care of it. It may not be in the library or mathlib, however.

I _think_ the newest version of Lean (or yesterday's, tonight's seems broken) should manage this just with `simp`.

**Comment by semorrison** [comment]
If you add a dependency on lean-tidy (via `leanpkg add git@github.com:semorrison/lean-tidy.git`) and then `import tidy.congr_struct`, you'll have the `congr_struct tactic` available.

An alternative is to prove a custom lemma, along the lines of https://github.com/semorrison/lean-category-theory/blob/master/src/categories/natural_transformation.lean#L28 in the category theory library.

**Comment by semorrison** [comment]
My Lean is up and running again. If you include a link to the place you're seeing this, I'll have a look.

**Comment by MitchRowett** [comment]
I managed to get this working using a lemma, but I'm now hitting the opposite problem. If I have a hypothesis `h` that two subgroups are equal, how would I use that to show that the underlying sets are equal?

I've also had a read of the gitter conversation this morning. I think I followed most of the ideas behind the alternative structure you and Mario were discussing, but I'm not quite sure whether I should implement it, or how I would go about it.
